deeplinkdispatch-base/src/main/java/com/airbnb/deeplinkdispatch/base/MatchIndex.java

```diff
@@ -145,26 +143,34 @@ private String compareValue(int elementStartPos, byte inboundUriComponentType, @
     inboundValue, Map<String, String> pathSegmentReplacements) {
     // Placeholder always matches
     int valueStartPos = elementStartPos + HEADER_LENGTH;
-    byte nodeMetadata = byteArray[elementStartPos];
-    boolean isComponentParam = isComponentParam(nodeMetadata);
-    boolean isConfigurablePathSegment = isConfigurablePathSegment(nodeMetadata);
-    boolean isValueLiteralValue = !(isComponentParam || isConfigurablePathSegment);
+    NodeMetadata nodeMetadata = new NodeMetadata(byteArray[elementStartPos]);
+
+    //Opportunistically skip doing more operations on this node if we can infer it will not match
+    //based on URIComponentType
+    if (nodeMetadata.isComponentTypeMismatch(inboundUriComponentType)) return null;
+
     int valueLength = getValueLength(elementStartPos);
-    boolean isComponentTypeMismatch = nodeMetadata != inboundUriComponentType;
     boolean isValueLengthMismatch = valueLength != inboundValue.length;
-
-    if ((isComponentTypeMismatch || isValueLengthMismatch) && isValueLiteralValue) {
-      //Opportunistically skip comparing this node before walking the two values if we can infer
-      // that they will not match.
+    if (isValueLengthMismatch && nodeMetadata.isValueLiteralValue) {
+      //Opportunistically skip this node if the comparator's lengths mismatch.
       return null;
     }
-    if (isComponentParam) {
+
+    if (nodeMetadata.isComponentParam) {
       return compareComponentParam(valueStartPos, valueLength, inboundValue);
-    } else if (isConfigurablePathSegment) {
+    } else if (nodeMetadata.isConfigurablePathSegment) {
       return compareConfigurablePathSegment(inboundValue, pathSegmentReplacements, valueStartPos,
         valueLength);
+    } else {
+      return compareValuesWalk(inboundValue, valueStartPos);
     }
-    return ""; // Matches but is no placeholder
+  }
+
+  private String compareValuesWalk(byte[] inboundValue, int valueStartPos) {
```

Comment:
I know it does not matter but we should walk on the length in the array.
They are the same here (as we check that before0 but if they are not (we might need to change that) we would walk out of bounds inside the array and that is not safe.